File: solutions/N-Queens-II/N-Queens-II.cpp

class Solution {
public:
    bool ok(vector<int> visit,int endRow)
    {
        for(int row=0;row<endRow;row++)
            if(visit[row]==visit[endRow] || abs(visit[row]-visit[endRow]) == endRow-row)
                return false;
        return true;
    }
    void dfs(vector<int> &visit,int n,int row,int &ans)
    {
        if(row==n)
        {
            ans++;
            return;
        }
        for(int col=0;col<n;col++)
        {
            visit[row]=col;
            if(ok(visit,row))  
                dfs(visit,n,row+1,ans);
        }
    }
    int totalNQueens(int n) {
        vector<int> visit(n,0);
        int ans=0;
        dfs(visit,n,0,ans);
        return ans;
    }
};


 

File: solutions/N-Queens-II/N-Queens-II.js
/**
 * @param {number} n
 * @return {number}
 */
var totalNQueens = function (n) {
  const visit = Array.from({ length: n }, () => 0);
  let ans = 0;
  const ok = (endRow) => {
    for (let row = 0; row < endRow; row++) {
      if (visit[row] === visit[endRow] || Math.abs(visit[row] - visit[endRow]) === endRow - row) {
        return false;
      }
    }
    return true;
  };
  const dfs = (row) => {
    if (row === n) {
      ans++;
      return;
    }
    for (let col = 0; col < n; col++) {
      visit[row] = col;
      if (ok(row)) {
        dfs(row + 1);
      }
    }
  };
  dfs(0);
  return ans;
};